cluster/v1alpha1/helpers.go

```diff
@@ -0,0 +1,388 @@
+package v1alpha1
+
+import (
+	"fmt"
+	"math"
+	"regexp"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/utils/clock"
+	clusterv1beta1 "open-cluster-management.io/api/cluster/v1beta1"
+)
+
+var RolloutClock = clock.Clock(clock.RealClock{})
+var maxTimeDuration = time.Duration(math.MaxInt64)
+
+type RolloutStatus int
+
+const (
+	// resource desired status is not applied yet
+	ToApply RolloutStatus = iota
+	// resource desired status is applied and last applied status is not updated
+	Progressing
+	// resource desired status is applied and last applied status is succeed
+	Succeed
+	// resource desired status is applied and last applied status is failed
+	Failed
+	// when the rollout status is progressing or failed and the status remains for longer than the timeout
+	// then the status will be set to timeout
+	TimeOut
+	// skip rollout on this cluster
+	Skip
+)
+
+// Return the rollout status on each managed cluster
+type ClusterRolloutStatusFunc func(clusterName string) ClusterRolloutStatus
+
+type ClusterRolloutStatus struct {
+	// cluster name
+	// required field
+	Name string
+	// cluster group key
+	// optional field
+	GroupKey clusterv1beta1.GroupKey
+	// rollout status
+	// required field
+	Status RolloutStatus
+	// the last transition time of rollout status
+	// this is used to calculate timeout for progressing and failed status
+	// optional field
+	LastTransitionTime *time.Time
+}
+
+type RolloutResult struct {
+	// clusters to rollout
+	ClustersToRollout map[string]ClusterRolloutStatus
```

Comment:
It seems redundant that `ClusterRolloutStatus` has a cluster name field when it's used in the context of a map where the keys are cluster names.

My opinion is to remove `ClusterRolloutStatus.Name` or make `ClustersToRollout` a slice of `[]ClusterRolloutStatus`.

The same feedback applies for `ClustersTimeOut`.

Reply:
Fixed.